crates/masonry/src/text2/edit.rs

```diff
@@ -101,38 +101,45 @@ impl<T: EditableText> TextEditor<T> {
                             Handled::No
                         }
                         Key::Named(NamedKey::Space) => {
-                            if let Some(selection) = self.selection.selection {
-                                // TODO: We know this is not the fullest model of copy-paste, and that we should work with the inner text
-                                // e.g. to put HTML code if supported by the rich text kind
-                                let c = ' ';
-                                self.text_mut().edit(selection.range(), c);
-                                self.selection.selection = Some(Selection::caret(
-                                    selection.min() + c.len_utf8(),
-                                    // We have just added this character, so we are "affined" with it
-                                    Affinity::Downstream,
-                                ));
-                                Handled::Yes
-                            } else {
-                                debug_panic!("Got text input event whilst not focused");
-                                Handled::No
-                            }
+                            let selection = self.selection.selection.unwrap_or(Selection {
+                                anchor: 0,
+                                active: 0,
+                                active_affinity: Affinity::Downstream,
+                                h_pos: None,
+                            });
+                            let c = ' ';
+                            self.text_mut().edit(selection.range(), c);
+                            self.selection.selection = Some(Selection::caret(
+                                selection.min() + c.len_utf8(),
+                                // We have just added this character, so we are "affined" with it
+                                Affinity::Downstream,
+                            ));
+                            let contents = self.text().as_str().to_string();
+                            ctx.submit_action(Action::TextChanged(contents));
```

Comment:
I was trying to avoid passing the context in here (e.g. because its origin is wrong). Not sure what the right interface here is, though.

Reply:
I think this code will be due for a refactor after RustNL anyway (not to be dismissive of your work, I just think we haven't decided on a lot of semantics yet).